[Blog] Grain State of the Union 2020 (#113)
grain-lang/grain-lang.org: src_blog/_posts/Grain-State-of-the-Union-2020.md

---
title: Grain State of the Union 2020
subtitle: Every update you need to know, and what's to come in 2021.
date: 2020-09-29 12:00:00
author: Oscar Spencer
cover: /blog/photos/cover/GrainStateOfTheUnionCover.jpg

> coverAttribution: Luke Chesser / Unsplash

Reviewer: Did you edit a photo by him with the logo on it?

Author: I suppose I could make my own gradient, but I was lazy.

## The Grain Mission

When we first started Grain in April 2017, WebAssembly had just been turned on by default in the four major browsers. There were very few languages that could compile to Wasm, and there weren't really any _new_ languages for WebAssembly. In fact, the premier language for WebAssembly at the time was C. We saw that there was a real opportunity to create a new language whose sole existence is to be a general-purpose, high level, accessible language for WebAssembly, and thus, Grain was born.

### What We're Building

We want to put academic language features to work for the everyday developer. As we're building this new language, we want to collect the very best that we've seen from other languages and present them as one unified, modern experience. This includes things like having a number system that's inspired by that of [Racket](https://racket-lang.org/), where rational numbers are first-class, or having efficient pattern matching that makes it easy to reason about your program.

> Grain aims to be what I would describe as a _relaxed_ functional language. This means that we'll still have a strong typechecker and still support most functional language constructs, but we'll also have features like `let mut` that steer a little off the beaten path for functional languages. These deviances are what we feel to be pragmatic and make the language more practical for everyday use.

Reviewer: ```suggestion
Grain aims to be what I would describe as a _relaxed_ functional language. This means that we'll still have a strong typechecker and still support most functional language constructs, but we'll also adopt features like [Rust's](https://www.rust-lang.org) `let mut` that steer a little off the beaten path for functional languages. These deviances are what we feel to be pragmatic and make the language more practical for everyday use.
```

Reviewer: I like the other language shoutouts, but I'm okay if you don't want this one.

Of course, no open source project would be anything without a strong community. Continuing to build a vibrant community will be a big part of what makes Grain successful. Our [Discord server](https://discord.gg/7U3newJ) has a great group of people and is growing to become something truly great. We hope to create a collaborative space where everyone feels welcome to ask, learn, and contribute.

## Core Language Development

We've been pretty busy as of late. As of writing, we've merged **157** pull requests in just the past year. It's crazy to realize how far the language has come in such a short amount of time, and I'm immensely proud of what we've accomplished. Here are some of the highlights.

### Language Features

> While we completed some of the less exciting but critically important features like garbage collection, we also revamped our import/export system, expanded our `Number` type, implemented more data structures, added mutable `let` bindings, added support for generating source maps, and more.

Reviewer: ```suggestion
We completed some of the less exciting, yet critically important, features like garbage collection. We also revamped our import/export system, expanded our `Number` type, implemented more data structures, added mutable `let` bindings, added support for generating source maps, and more!
```

We've also continued to buff out the rough edges around the language, making Grain code more pleasant to write.

### Development Workflow

Much of our progress was made possible by making it easier for newcomers (and us!) to contribute to the project. This included a whole new build process using familiar tools more people already had on their systems, and simpler instructions to get up and running. We also decided to make one pretty big switch—the compiler is now written in ReasonML rather than OCaml. While this is largely an aesthetic change, we've found that contributors are more comfortable with the Reason syntax and that it's a bit less intimidating for those who are unfamiliar with the OCaml/Reason ecosystem.

> We've also made writing the compiler code easier—we now use [Binaryen](https://github.com/WebAssembly/binaryen) to generate our WebAssembly instructions via a new project we made called [Binaryen.ml](https://github.com/grain-lang/binaryen.ml). This makes it far easier to reliably generate programs, rather than writing out every instruction by hand. On that same front, we now use [AssemblyScript](https://www.assemblyscript.org/) for the more complex operations necessary for some of the built-in standard library functions, and that experience has been delightful.

Reviewer: ```suggestion
We've also made writing the compiler code easier—we now use [Binaryen](https://github.com/WebAssembly/binaryen) to generate our WebAssembly instructions via a new project we made called [Binaryen.ml](https://github.com/grain-lang/binaryen.ml). This makes it far easier to reliably generate programs, rather than writing out every instruction by hand. On that same front, we now use [AssemblyScript](https://www.assemblyscript.org/) for the lower-level operations necessary for some of the built-in standard library functions, and that experience has been delightful.
```

Reviewer: Seems weird to say Grain can't handle "complex" things, instead what we are really saying is like manual memory management and stuff, so I switched it to "lower-level"

### Contributors

We've welcomed [Blaine Bublitz](https://twitter.com/BlaineBublitz) to the Grain core team! 🎉

> Blaine has brought his exceptional experience maintaining open source projects to Grain and has truly been the catalyst that's gotten us to where we are today. I'm not really sure how we ever functioned without him.

Reviewer: 😊

### Logo & Website Redesign

![grain redesign](/blog/photos/misc/grain-redesign.png)

If you've followed Grain from the beginning, you might remember this old logo. While this logo (hastily-created in about five minutes by yours truly) served us well, we've moved on to our new, beautiful, professionally-designed logo that I feel fits us wonderfully.

We've also got a brand new website! Given that this is an open source project, all of the budget went to the logo and the website design was left to me. 😅 It's been well-received, but we're always open to feedback from the community.

### The Grain Guide

In addition to the regular docs, we've launched the Grain Guide. The purpose of the Guide is to teach the reader how to use the language without overwhelming them with everything Grain can do. It introduces them to new concepts and patterns as they read on, letting them check out the normative documentation as they see fit.

## Our Roadmap

The language has matured a ton, but there's always lots more to do.

> As for big features, we're continuing to closely monitor the WebAssembly [exception handling proposal](https://github.com/WebAssembly/exception-handling) to bring better exception handling to Grain. In the meantime, we're working on support for macros—stay tuned for more news. As for smaller (but just as exciting) features, we're continuing to build out our standard library (including the DOM library rewrite!), and improve the general ergonomics of the language.

Reviewer: ```suggestion
As for big features, we're continuing to closely monitor the WebAssembly [exception handling proposal](https://github.com/WebAssembly/exception-handling) to bring better exception handling to Grain. In the meantime, we're discussing macro support—stay tuned for more news. As for smaller (but just as exciting) features, we're continuing to build out our standard library (including the DOM library rewrite!), and improve the general ergonomics of the language.
```
